app/controllers/jobseekers/sessions_controller.rb

```diff
@@ -1,4 +1,14 @@
 class Jobseekers::SessionsController < Devise::SessionsController
   after_action :replace_devise_notice_flash_with_success!, only: %i[create destroy]
   before_action :sign_out_publisher!, only: %i[create]
+
+  def new
```

Comment:
how about doing this instead:

```
  def replace_devise_error_flash_with_form_errors
    return unless AUTHENTICATION_FAILURE_MESSAGES.include?(flash[:alert])

    message = flash[:alert]
    flash.clear
    render_resource_with_error(:new, :email, message)
  end
```

You would need to move the method `render_resource_with_error` from `RegistrationsController` to `ApplicationController` and take as parameter the view as well

Reply:
There's two problems with this, but it's a helpful idea that will help me come up with a better solution.

The main problem is that the before_action skips the form validation errors - it immediately checks whether Devise could authenticate and renders if it can't.

Secondly, I can't use resource_class here. I have to use a form object, in order to keep separate the different validations on the sign in / sign up pages (e.g. non-unique email address). But that problem can easily be solved by writing a `render_form_with_error` method.

How about this?

```ruby
def new
    @sign_in_form = JobseekerSignInForm.new(sign_in_params)

    if params[:action] == "create" && @sign_in_form.invalid?
      clear_flash_and_render(:new)
    elsif AUTHENTICATION_FAILURE_MESSAGES.include?(flash[:alert])
      message = flash[:alert]
      @sign_in_form.errors.add(:email, message)
      clear_flash_and_render(:new)
    else
      super
    end
  end

private

  def clear_flash_and_render(view)
    flash.clear
    render view
  end
```